### ISSUE TITLE
New test for Factors

### ISSUE THREAD

**Issue by khalibartan** [pgmpy/pgmpy#590]
I was thinking of rewriting the test for Factor class.
- [ ] I'll separate the tests for Factor, TabularCPD and JointProbabilityDistribution classes and will try to add more tests for Factors and JointProbabilityDistribution (not for CPD because @ankurankan is looking forward to remove some redundant properties from it).
- [ ] I'll test cases where variables can be a list hashable objects like tuples etc.
- [ ] I hope to write tests that which checks the boundary conditions for the methods.
- [ ] Coverage of the current skipping test for reduce method.

@ankurankan Should I do the following things or not.
Other people can also add in there views, related to tests.

**Comment by mvnnn** [comment]
@khalibartan  i am new contributor. i like to work on this. should i do ?

**Comment by khalibartan** [comment]
@mvnnn This is not an easy issue in my view, actually I wanted to work on this one, I haven't discuss things with @ankurankan .

**Comment by ankurankan** [comment]
@khalibartan You can go ahead with this. We need to clean up a lot of tests. Would be a good start.

**Comment by khalibartan** [comment]
@ankurankan Will send a PR ASAP.

**Comment by khalibartan** [comment]
Would someone like to work on it. It looks like I won't be able to give necessary time because of my mid-term examinations. I would be great if someone could finish it up. Otherwise I'll work on it after my mid-terms are over.

**Comment by ankurankan** [comment]
Hi Developers,

I am a newbie here but would like to give this a shot. Can someone please
elaborate a bit on the issue?

Thanks

**Comment by khalibartan** [comment]
What we are trying to do is have new test for our Factor class objects. The current tests are not testing things to the boundary conditions. Also test don't test the factor class , where scope can be a list of immutable objects.

**Comment by abhilashrj** [comment]
@anaviltripathi Are u working on this right now?

**Comment by ankurankan** [comment]
@abhi95 I don't think anyone is working on this right now. So, you can go ahead with this if you like.

**Comment by anaviltripathi** [comment]
@abhi95 No, I am not working on this. You can go ahead with, and make sure to include some asymmetric cases whenever you write tests. Currently many tests use symmetric factors/CPDs.

**Comment by abhilashrj** [comment]
@anaviltripathi by asymmetric factors do you mean variables with different cardinality

**Comment by anaviltripathi** [comment]
@abhi95 Yes, and different values too at different places in the table. Some of the current tests have the same set of numbers for all the places.

**Comment by abhilashrj** [comment]
@anaviltripathi Regarding variables being tuples instead of lists, these tests are just to verify that methods are not altering provided data right. What other immutable objects I can consider while writing tests?

**Comment by anaviltripathi** [comment]
@abhi95 Which tests are you talking about in particular?

**Comment by khalibartan** [comment]
@abhi95 Variables name can be any hashable object, by mistake I might have written immutable.

**Comment by abhilashrj** [comment]
@anaviltripathi I was talking about variables being hashable objects such as numbers, strings, tuples, etc. anyway @khalibartan cleared my doubt. I think you were right about immutable objects, as only built in immutable objects are hashable in python.
